### ISSUE TITLE
Default value Handling not working for JSON object and array

### ISSUE THREAD

**Issue by arukris-cdw** [strongloop/loopback-connector-mongodb#552]
hi Team, 

This issue is in continuation to the solution shared for handling default values in below thread.
https://github.com/strongloop/loopback-connector-mongodb/issues/534

Looks like the fix is working for elements only at the root level, but not for elements inside an object / object array. 
 
Attached is sample input, output and model json I prepared for this testing. Could you please review and let me know if I am missing something in the configuration. ?.
 
[Input_and_output.txt](https://github.com/strongloop/loopback-connector-mongodb/files/3859420/Input_and_output.txt)

[quotes_model.txt](https://github.com/strongloop/loopback-connector-mongodb/files/3859429/quotes_model.txt)

**below is the version details.** 

C:\Projects\mongodb\mongocloud>npm ls loopback-datasource-juggler
mongocloud@1.0.0 C:\Projects\mongodb\mongocloud
+-- loopback@3.26.0
| +-- loopback-connector-remote@3.4.1
| | `-- loopback-datasource-juggler@3.33.2
| `-- loopback-datasource-juggler@3.33.2
`-- loopback-datasource-juggler@4.14.0

**Comment by dhmlau** [comment]
@hacksparrow, I'm assigning this to you. Thanks!

**Comment by hacksparrow** [comment]
Hi @arukris-cdw we now have a PR for this support - https://github.com/strongloop/loopback-datasource-juggler/pull/1797.

@bajtos @raymondfeng PTAL.

**Comment by arukris-cdw** [comment]
@hacksparrow 
Thanks. Please let me know what version to install to test this feature..

**Comment by hacksparrow** [comment]
@arukris-cdw will do once we land the PR and publish and new version.

**Comment by hacksparrow** [comment]
@arukris-cdw `loopback-datasource-juggler 4.16.0` has been published with the supported feature.

**Comment by dhmlau** [comment]
Closing as done.

**Comment by arukris-cdw** [comment]
@hacksparrow , 

this issue still exists. below is my versions and attached my quotes.json file. Can you please review and let me know if everything looks good. 


C:\Projects\mongodb\mongocloud>npm install loopback-datasource-juggler
+ loopback-datasource-juggler@4.16.0
updated 1 package in 6.984s

C:\Projects\mongodb\mongocloud>npm ls loopback-datasource-juggler
mongocloud@1.0.0 C:\Projects\mongodb\mongocloud
+-- loopback@3.26.0
| +-- loopback-connector-remote@3.4.1
| | `-- loopback-datasource-juggler@3.33.2
| `-- loopback-datasource-juggler@3.33.2
`-- loopback-datasource-juggler@4.16.0

[quotes_model.txt](https://github.com/strongloop/loopback-connector-mongodb/files/3917848/quotes_model.txt)

**Comment by arukris-cdw** [comment]
[mongocloud.zip](https://github.com/strongloop/loopback-connector-mongodb/files/3932899/mongocloud.zip)

@hacksparrow , i have uploaded the zip file of the application .

**Comment by hacksparrow** [comment]
@arukris-cdw default values work in two ways:

1. Getting written to the database if the value was not specified
2. Getting applied to the response IF the data was missing in the database

`applyDefaultOnWrites` is a property-level setting which controls whether the default value is written to the database or not.

`applyDefaultsOnReads` is a model property which controls whether the default should be applied to the response object or not.

In your case, the defaults are actually not written to the database. However, since `applyDefaultsOnReads` defaults to `true`, the defaults are being applied to the response objects.

To stop LoopBack from applying the defaults in the response object, just set `applyDefaultsOnReads` to `false`. Example:

```js
{
...
	"name": "movies",
	"type": "object",
	"title": "The movies Schema",
	"options": {
		"validate": false,
		"forceId": false
	},
	"applyDefaultsOnReads": false,
	"properties": {
		"title": {
			"type": "string"
		},
...
```
Hope it helps.

**Comment by dhmlau** [comment]
@arukris-cdw, 

There are 2 fixes from @hacksparrow 
3.x branch: https://github.com/strongloop/loopback-datasource-juggler/pull/1815
master branch: https://github.com/strongloop/loopback-datasource-juggler/pull/1813
Both have landed. 

I've released `loopback-datasource-juggler@3.35.0` for the fixes in 3.x branch.  Please give it a try. 
I published a release for `master` because I think @arukris-cdw is only waiting for fixes on 3.x, and I'd like to wait for more PRs to go into master before making another release.  

If you need it for `master`, please let me know and I can release it for 4.x. Thanks!

**Comment by dhmlau** [comment]
I'm closing this issue since @arukris-cdw has verified that it's working in a LB standalone application. Closing as resolved.
